src/java/org/apache/cassandra/locator/ReplicaPlans.java

```diff
@@ -593,8 +593,11 @@ public static ReplicaPlan.ForRangeRead forSingleReplicaRead(Keyspace keyspace, A
     public static ReplicaPlan.ForTokenRead forRead(Keyspace keyspace, Token token, Index.QueryPlan indexQueryPlan, ConsistencyLevel consistencyLevel, SpeculativeRetryPolicy retry)
     {
         AbstractReplicationStrategy replicationStrategy = keyspace.getReplicationStrategy();
-        EndpointsForToken candidates = candidatesForRead(keyspace, indexQueryPlan, consistencyLevel, ReplicaLayout.forTokenReadLiveSorted(replicationStrategy, token).natural());
-        EndpointsForToken contacts = contactForRead(replicationStrategy, consistencyLevel, retry.equals(AlwaysSpeculativeRetryPolicy.INSTANCE), candidates);
+        IEndpointSnitch endpointSnitch = DatabaseDescriptor.getEndpointSnitch();
+        EndpointsForToken candidates = candidatesForRead(keyspace, indexQueryPlan, consistencyLevel, ReplicaLayout.forTokenReadLiveSorted(replicationStrategy, token).natural())
+                                       .filter(endpointSnitch.filterByAffinity(keyspace.getName()));
+        EndpointsForToken contacts = contactForRead(replicationStrategy, consistencyLevel, retry.equals(AlwaysSpeculativeRetryPolicy.INSTANCE), candidates)
+                                     .filter(endpointSnitch.filterByAffinity(keyspace.getName()));
```

Comment:
This is the best option I could find for introducing `filterByAffinity`.

`ReplicaPlans.forRead` is called by `AbstractReadExecutor.getReadExecutor` (it's also the only production call to this method) and this corresponds to the DSE change which uses `filterByAffinity` to modify the `scratchAllReplicas` list.